pypyr/loaders/string.py

```diff
@@ -0,0 +1,32 @@
+import logging
+from functools import wraps
+
+from pypyr.cache.stringcache import cache
+from pypyr.pipedef import PipelineDefinition, PipelineFileInfo
+from pypyr.yaml import get_pipeline_yaml
+
+logger = logging.getLogger(__name__)
+
+
+def log(func):
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        logger.debug("starting")
+        result = func(*args, **kwargs)
+        logger.debug("done")
+        return result
+
+    return wrapper
+
+
+@log
+def load_pipeline(pipeline, parent):
+    info = PipelineFileInfo(pipeline_name="", parent=parent, loader=__name__, path=None)
+    return PipelineDefinition(pipeline=get_pipeline_yaml(pipeline), info=info)
+
+
+@log
+def get_pipeline_definition(pipeline_name, parent):
+    definition = cache.get(pipeline_name, lambda: load_pipeline(pipeline_name, parent))
```

Comment:
great instinct to want to cache this - the yaml parsing is indeed expensive!

but, so happens pypyr actually already [caches the PipelineDefinition](https://github.com/pypyr/pypyr/blob/bc1a53043700738b1269b2f8a5ff019ec818a676/pypyr/cache/loadercache.py#L70) in the loader cache (which loads the loader you're writing here will also actually keep a cache of PipelineDefinitions loaded by that loader).

So this is actually just duplicating memory use (well, potentially, depending on whether the strings get interned).


You could instead simplify this to:

```python
 @log
 def get_pipeline_definition(pipeline_name, parent):
     definition = load_pipeline(pipeline_name, parent)
     logger.debug("found %d stages in pipeline.", len(definition.pipeline))
     return definition
```

And take advantage of the loader_cache caching the PipelineDefinition for you without you having to do anything extra. This means you can then remove `pypyr/cache/stringcache.py` entirely.

Reply:
Done in ebc81ec